File: app/src/main/java/com/hezb/hplayer/util/PlayerGesture.java
package com.hezb.hplayer.util;

import android.app.Activity;
import android.graphics.Color;
import android.os.Handler;
import android.util.Pair;
import android.util.TypedValue;
import android.view.GestureDetector.OnGestureListener;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.hezb.hplayer.R;
import com.hezb.hplayer.ui.view.LightnessVolumeView;

import io.vov.vitamio.utils.ScreenResolution;
import io.vov.vitamio.utils.StringUtils;
import io.vov.vitamio.widget.VideoView;

/**
 * 播放器手势操作
 * TODO 需优化 问题不少！
 */
public class PlayerGesture implements OnGestureListener {

    private VideoView mVideoView;
    private LightnessVolumeView mLV;// 音量，亮度
    private TextView mTime;// 快进/快退 时间
    private OnSingleTapUpListener singleTapUpListener;
    private Handler mHandler;

    private int left; // VideoView 左边坐标
    private int right; // VideoView 右边坐标
    private int width;// VideoView 宽
    private int currentSeek = -1;// 当前需要跳转的进度
    private boolean intoSeek = false;// 进入 快进/快退 状态

    public PlayerGesture(Activity activity, VideoView videoView, RelativeLayout rootView) {
        mVideoView = videoView;

        Pair<Integer, Integer> screenPair = ScreenResolution.getResolution(activity);
        width = screenPair.first;
        left = 0;
        right = width;

        // 添加音量，亮度设置展示View，VideoView 的父布局必须是 RelativeLayout
        RelativeLayout.LayoutParams lp = new RelativeLayout.LayoutParams(
                DensityUtil.dip2px(activity, 60), DensityUtil.dip2px(activity, 60));
        lp.addRule(RelativeLayout.CENTER_IN_PARENT);
        mLV = new LightnessVolumeView(activity);
        mLV.setLayoutParams(lp);
        mLV.setActivity(activity);
        mLV.setBackgroundResource(R.drawable.loading_bg);
        mLV.setGravity(Gravity.CENTER);
        mLV.setVisibility(View.GONE);
        rootView.addView(mLV);

        RelativeLayout.LayoutParams lp2 = new RelativeLayout.LayoutParams(
                RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
        lp2.addRule(RelativeLayout.CENTER_IN_PARENT);
        mTime = new TextView(activity);
        mTime.setLayoutParams(lp2);
        mTime.setTextSize(TypedValue.COMPLEX_UNIT_SP, 16);
        mTime.setTextColor(Color.parseColor("#ffffff"));
        mTime.setGravity(Gravity.CENTER);
        mTime.setBackgroundResource(R.drawable.loading_bg);
        mTime.setVisibility(View.GONE);
        rootView.addView(mTime);
        mTime.setPadding(20, 10, 20, 10);

        mHandler = new Handler();
    }

    public void setOnSingleTapUpListener(OnSingleTapUpListener listener) {
        singleTapUpListener = listener;
    }

    @Override
    public boolean onDown(MotionEvent e) {
        return false;
    }

    @Override
    public void onShowPress(MotionEvent e) {
    }

    @Override
    public boolean onSingleTapUp(MotionEvent e) {
        if (singleTapUpListener != null) {
            return singleTapUpListener.onSingleTapUp(e);
        }
        return false;
    }

    @Override
    public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX,
                            float distanceY) {
        float x1 = e1.getX();
        float y1 = e1.getY();
        float x2 = e2.getX();
        float y2 = e2.getY();
        float absX = Math.abs(x1 - x2);
        float absY = Math.abs(y1 - y2);
        float absDistanceX = Math.abs(distanceX);// distanceX < 0 从左到右
        float absDistanceY = Math.abs(distanceY);// distanceY < 0 从上到下
        if (absDistanceX < absDistanceY && !intoSeek) { // Y方向的速率比X方向的大，即 上下 滑动
            if (distanceY > 0) {
                if (x1 >= (right - width * 0.35)) {
                    //在右边，调节声音
                    mLV.changeVolume(LightnessVolumeView.ADD_FLAG);
                } else if (x1 <= (left + width * 0.35)) {
                    //在左边，调节亮度
                    mLV.changeLightness(LightnessVolumeView.ADD_FLAG);
                }

                return true;
            } else {
                if (x1 >= (right - width * 0.35)) {
                    //在右边，调节声音
                    mLV.changeVolume(LightnessVolumeView.SUB_FLAG);
                } else if (x1 <= (left + width * 0.35)) {
                    //在左边，调节亮度
                    mLV.changeLightness(LightnessVolumeView.SUB_FLAG);
                }

                return true;
            }

        } else { // 左右
            if (absX > absY) {
                intoSeek = true;
                float progress = absX / width * 1.2f;
                if (x1 > x2) {
                    onRewind(progress);
                } else {
                    onForward(progress);
                }
                return true;
            }
        }

        return false;
    }

    @Override
    public void onLongPress(MotionEvent e) {
    }

    @Override
    public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX,
                           float velocityY) {
        return false;
    }

    private void onForward(float progress) {
        int seek = (int) (mVideoView.getCurrentPosition()
                + mVideoView.getDuration() * progress);
        if (seek < mVideoView.getDuration()) {
            showTime(seek);
        }
    }

    private void onRewind(float progress) {
        int seek = (int) (mVideoView.getCurrentPosition()
                - mVideoView.getDuration() * progress);
        if (seek >= 0) {
            showTime(seek);
        }
    }

    private void showTime(int seek) {
        currentSeek = seek;
        String showTime = StringUtils.generateTime(seek) + "/" + StringUtils.generateTime(mVideoView.getDuration());
        mTime.setText(showTime);
        if (mTime.getVisibility() != View.VISIBLE) {
            mTime.setVisibility(View.VISIBLE);
        }
    }

    /**
     * 手指抬起
     */
    public void onFingerUp() {
        if (currentSeek != -1) {
            mVideoView.seekTo(currentSeek);
            mHandler.postDelayed(hideTimeView, 500);
        }
        currentSeek = -1;
        intoSeek = false;
    }

    private Runnable hideTimeView = new Runnable() {
        @Override
        public void run() {
            mTime.setVisibility(View.GONE);
        }
    };


    public interface OnSingleTapUpListener {
        boolean onSingleTapUp(MotionEvent e);
    }
}

File: playerCore/src/main/java/com/hezb/player/gesture/PlayerGesture.kt
package com.hezb.player.gesture

import android.content.Context
import android.view.GestureDetector
import android.view.MotionEvent
import kotlin.math.abs

/**
 * Project Name: HPlayer
 * File Name:    PlayerGesture
 *
 * Description: 手势.
 *
 * @author  hezhubo
 * @date    2022年03月02日 17:12
 */
class PlayerGesture(context: Context) : GestureDetector.OnGestureListener {

    /** 左右两边的上下滑动范围监听比例，最大0.5f */
    private val LEFT_RIGHT_LIMIT = 0.4f
    /** 手势检测器 */
    private val mGestureDetector = GestureDetector(context.applicationContext, this)
    /** 单击监听器 */
    private var mSingleTapUpListener: OnSingleTapUpListener? = null
    /** 手势滑动监听器 */
    private var mOnScrollListener: OnScrollListener? = null
    /** 触摸控件的宽度 */
    private var width = 0
    /** 是否为seek状态 */
    private var isSeekState = false
    /** 是否是否需要左右滑动进行seek */
    private var needSeek = true

    /**
     * 设置是否需要快进快退手势
     *
     * @param needSeek
     */
    fun needLeftRightSeek(needSeek: Boolean) {
        this.needSeek = needSeek
    }

    /**
     * 设置单击监听器
     * 必须设置手势滑动监听器此监听器才生效
     *
     * @param l
     */
    fun setOnSingleTapUpListener(l: OnSingleTapUpListener) {
        mSingleTapUpListener = l
    }

    /**
     * 设置手势滑动监听器
     *
     * @param l
     */
    fun setOnScrollListener(l: OnScrollListener) {
        mOnScrollListener = l
    }

    /**
     * 触摸事件
     *
     * @param event
     * @param viewWidth
     * @return
     */
    fun onTouchEvent(event: MotionEvent, viewWidth: Int): Boolean {
        width = viewWidth
        if (mOnScrollListener != null) {
            val action = event.action
            if (action == MotionEvent.ACTION_DOWN) {
                onFingerDown()
            } else if (action == MotionEvent.ACTION_UP || action == MotionEvent.ACTION_CANCEL) {
                onFingerUp()
            }
            mGestureDetector.onTouchEvent(event)
            return true
        }
        return false
    }

    override fun onShowPress(e: MotionEvent) {
    }

    override fun onSingleTapUp(e: MotionEvent): Boolean {
        return mSingleTapUpListener?.onSingleTapUp(e) ?: false
    }

    override fun onDown(e: MotionEvent): Boolean {
        return false
    }

    override fun onFling(
        e1: MotionEvent,
        e2: MotionEvent,
        velocityX: Float,
        velocityY: Float
    ): Boolean {
        return false
    }

    override fun onScroll(
        e1: MotionEvent,
        e2: MotionEvent,
        distanceX: Float,
        distanceY: Float
    ): Boolean {
        mOnScrollListener?.let {
            val x1 = e1.x
            val y1 = e1.y
            val x2 = e2.x
            val y2 = e2.y
            val absX = abs(x1 - x2)
            val absY = abs(y1 - y2)
            val absDistanceX = abs(distanceX) // distanceX < 0 从左到右
            val absDistanceY = abs(distanceY) // distanceY < 0 从上到下
            if (absDistanceX < absDistanceY && !isSeekState) { // Y方向的速率比X方向的大，即 上下 滑动
                return if (distanceY > 0) { // 增加
                    if (x1 >= width * (1 - LEFT_RIGHT_LIMIT)) { // 在右边
                        it.onRightSideUpDown(true)
                    } else if (x1 <= width * LEFT_RIGHT_LIMIT) { //在左边
                        it.onLeftSideUpDown(true)
                    }
                    true
                } else {
                    if (x1 >= width * (1 - LEFT_RIGHT_LIMIT)) { //在右边
                        it.onRightSideUpDown(false)
                    } else if (x1 <= width * LEFT_RIGHT_LIMIT) { //在左边
                        it.onLeftSideUpDown(false)
                    }
                    true
                }
            } else if (needSeek) { // 左右
                if (absX > absY) {
                    isSeekState = true
                    it.onScrollLeftRight(width, distanceX)
                    return true
                }
            }
        }
        return false
    }

    override fun onLongPress(e: MotionEvent?) {
    }

    /**
     * 单个手指按下
     */
    private fun onFingerDown() {
        mOnScrollListener?.onFingerDown()
    }

    /**
     * 单个手指抬起
     */
    fun onFingerUp() {
        mOnScrollListener?.onFingerUp()
        isSeekState = false
    }

    /**
     * 单击事件监听器
     */
    interface OnSingleTapUpListener {
        fun onSingleTapUp(e: MotionEvent): Boolean
    }

    /**
     * 滑动监听器
     */
    interface OnScrollListener {
        /**
         * 在左边上下滑动
         *
         * @param isUp 是否向上
         */
        fun onLeftSideUpDown(isUp: Boolean)

        /**
         * 在右边
         *
         * @param isUp 是否向上
         */
        fun onRightSideUpDown(isUp: Boolean)

        /**
         * 在左右滑动滑动
         * 需要在手指抬起时叠加计算，再进行seek
         * 若直接调用seek，则对播放器操作十分频繁
         *
         * @param width    触摸控件的宽度
         * @param distance 滑动距离  向左>0  向右<0
         */
        fun onScrollLeftRight(width: Int, distance: Float)

        /**
         * 手指按下
         */
        fun onFingerDown()

        /**
         * 手指抬起
         */
        fun onFingerUp()
    }
}